File: dotnet/Multiplayer/ApiResponse.cs
using Michitai.Multiplayer.Errors;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Michitai.Multiplayer
{
    /// <summary>
    /// Base class for all API responses. Provides success status and error information.
    /// </summary>
    public abstract class ApiResponse
    {
        /// <summary>
        /// Indicates whether the API request was successful.
        /// </summary>
        public bool Success { get; set; }

        /// <summary>
        /// Error message if the request failed, null otherwise.
        /// </summary>
        public string? Error { get; set; }
    }

    /// <summary>
    /// Generic base class for API responses with typed error handling.
    /// Converts error strings to typed enums for better error management.
    /// </summary>
    /// <typeparam name="TError">The error enum type, must be an Enum and IConvertible.</typeparam>
    public abstract class ApiResponse<TError> : ApiResponse where TError : Enum, IConvertible
    {
        /// <summary>
        /// Gets the error type as an enum by converting the Error string.
        /// </summary>
        public TError ErrorType => ErrorConverter.ConvertToEnum<TError>(Error ?? string.Empty);

        /// <summary>
        /// Gets a user-friendly error message based on the error type.
        /// </summary>
        public string ErrorMessage => ErrorConverter.GetErrorMessage(ErrorType);
    }
}

File: python/multiplayer/api_response.py
"""Base classes for all API responses."""

from __future__ import annotations

import enum
from datetime import datetime
from dataclasses import dataclass, field
from typing import Any, ClassVar, Optional, Type

from .errors.converter import convert_to_enum, get_error_message
from .errors.enums import CommonError


@dataclass
class BanInfo:
    """Ban details returned when a banned player calls an endpoint."""

    ban_id: str = ""
    ban_duration: str = ""
    ban_reason: Optional[str] = None
    banned_at: Optional[datetime] = None
    banned_until: Optional[datetime] = None


@dataclass
class ApiResponse:
    """Base class for every API response."""

    #: The error enum type used by :attr:`error_type`. Subclasses override this.
    error_enum: ClassVar[Type[enum.Enum]] = CommonError

    success: bool = False
    error: Optional[str] = None
    ban_info: Optional[BanInfo] = field(default=None)

    @property
    def error_type(self) -> enum.Enum:
        """The typed error for :attr:`error`, or ``Unknown`` when unmapped."""
        return convert_to_enum(self.error, self.error_enum)

    @property
    def error_message(self) -> str:
        """A user-friendly message for :attr:`error_type`."""
        return get_error_message(self.error_type)


@dataclass
class SuccessResponse(ApiResponse):
    """Standard success response for operations that return no specific data."""

    message: str = ""
    updated_at: Optional[datetime] = None
